### ISSUE TITLE
MissingSwitchDefault false positive for deconstruction pattern in switch statement

### ISSUE THREAD

**Issue by kenwang-coursera** [checkstyle/checkstyle#15538]
I have read check documentation: https://checkstyle.org/checks/xxxxxx/nameofaffectedcheck.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
➜  Downloads javac Example.java                                                                                                            
➜  Downloads cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="MissingSwitchDefault"/>
    </module>
</module>
➜  Downloads cat Example.java 
package org.example;

public class Example {

  sealed interface InterfaceWithSealed permits ClassWithSealedA, ClassWithSealedB {}

  public record ClassWithSealedA() implements InterfaceWithSealed {}

  public record ClassWithSealedB() implements InterfaceWithSealed {}

  public static void main(String[] args) {
    final InterfaceWithSealed object = new ClassWithSealedA();
    switch (object) {
      case ClassWithSealedA() -> System.out.println("ClassWithSealedA");
      case ClassWithSealedB() -> System.out.println("ClassWithSealedA");
    }
  }
}

➜  Downloads RUN_LOCALE="-Duser.language=en -Duser.country=US"
➜  Downloads java $RUN_LOCALE -jar checkstyle-10.17.0-all.jar -c config.xml Example.java 
Starting audit...
[ERROR] /Users/kenwang/Downloads/Example.java:13:5: switch without "default" clause. [MissingSwitchDefault]
Audit done.
Checkstyle ends with 1 errors.
```

---

When using a switch **statement** with the deconstruction pattern, checkstyle reports a false positive for missing "default" clause, which is unnecessary due to the compiler's exhaustive matching for this pattern, and thus should be skipped in accordance to the [documentation](https://checkstyle.sourceforge.io/checks/coding/missingswitchdefault.html) on MissingSwitchDefault.

**Comment by rnveach** [comment]
@kenwang-coursera Is this related to Issue #15504 ?

@mahfouz72 Feel free to take a look.

**Comment by kenwang-coursera** [comment]
It could be related, but seems different enough to me:
1. sealed record vs enum
2. https://github.com/checkstyle/checkstyle/issues/15504 appears to be due to switch variable being wrapped, but this issue seems to be related to the deconstruction pattern in the individual case

**Comment by kenwang-coursera** [comment]
Incidentally, it seems like mixing deconstruction with regular pattern matching does _not_ trigger the false positive.

```
➜  Downloads cat Example.java                                                           
package org.example;

public class Example {

  sealed interface InterfaceWithSealed permits ClassWithSealedA, ClassWithSealedB {}

  public record ClassWithSealedA() implements InterfaceWithSealed {}

  public record ClassWithSealedB() implements InterfaceWithSealed {}

  public static void main(String[] args) {
    final InterfaceWithSealed object = new ClassWithSealedA();
    switch (object) {
      case ClassWithSealedA() -> System.out.println("ClassWithSealedA");
      case ClassWithSealedB b -> System.out.println("ClassWithSealedA");
    }
  }
}

➜  Downloads java $RUN_LOCALE -jar checkstyle-10.17.0-all.jar -c config.xml Example.java
Starting audit...
Audit done.
```

**Comment by mahfouz72** [comment]
@kenwang-coursera  This should be related to https://github.com/checkstyle/checkstyle/issues/14891 
(record pattern in case label)

Fix is already merged and this issue will be fixed in next release.

**Comment by romani** [comment]
Release will be on this weekend.

**Comment by mahfouz72** [comment]
@kenwang-coursera  We got the release. please close this or let us know if there is anything unexpected.
